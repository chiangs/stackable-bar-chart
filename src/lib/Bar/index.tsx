import React, { useState, useEffect } from "react";
import type { BarData, BarProps } from "../__types";

type Props = BarProps;

const NAME_COMPONENT = "stackable-bar";

const Bar: React.FC<Props> = ({
  value = 0,
  label = "",
  color = undefined,
  background = undefined,
  portion = 0,
  renderPortion = 0,
  mode = "stacked",
  showPercentage = false,
  showTooltip = false,
  revealTooltipHandler = (data: BarData | null) => data,
}) => {
  const displayValue = showPercentage ? `${portion}%` : `${value}`;
  const valueWidth = mode === "stacked" ? renderPortion : portion;
  const tooltip = showTooltip ? "tooltip" : "";
  const transitionAnimation = "width .2s ease-out";

  // Default internal width
  const [barWidth, setBarWidth] = useState(0);
  // On load, update width
  useEffect(() => {
    setBarWidth(valueWidth);
  }, []);

  // Bar style
  const width = `${barWidth}%`;
  const styleBar = {
    width,
    background: color,
  };

  // Animation style
  const animatedStyle = {
    ...styleBar,
    transition: transitionAnimation,
  };

  // Set bar style
  const styleContainer = mode === "stacked" ? animatedStyle : styleBar;

  // Set bar text styles
  const styleText = {
    color: background,
  };
  // Create bar content: Stacked
  let barContent = (
    <span className="bar-value-text" style={styleText}>
      {value}
    </span>
  );

  if (mode === "linear") {
    barContent = (
      <div className="stacked-content" style={styleText}>
        <div className="bar-value-text">
          <h3>{displayValue}</h3>
        </div>
        <div className="bar-label-text">
          <p>
            {label}
            {showPercentage ? <span>&nbsp;-&nbsp;{value}</span> : null}
          </p>
        </div>
      </div>
    );
  }

  return (
    <div
      className={`${NAME_COMPONENT} ${mode} ${tooltip}`}
      style={styleContainer}
      data-testid={NAME_COMPONENT}
      onMouseOver={() =>
        revealTooltipHandler({
          label,
          value,
          color: background,
          background: color,
          percentage: portion,
        })
      }
      onMouseOut={() => revealTooltipHandler(null)}
    >
      <div className="bar-value" style={styleContainer}>
        {barContent}
      </div>
    </div>
  );
};

export default Bar;
