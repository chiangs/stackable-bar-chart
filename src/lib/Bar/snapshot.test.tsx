import * as React from "react";
import renderer from "react-test-renderer";
import { describe, expect, test } from "vitest";
import type { BarData, BarProps } from "../__types";
import Bar from ".";

describe("Bar", () => {
  test("Bar component renders correctly", () => {
    const props: BarProps = {
      value: 1,
      label: "test",
      // color: undefined,
      // background: undefined,
      // portion: 0,
      // renderPortion: 0,
      // mode: "stacked",
      // showPercentage: false,
      // showTooltip: false,
      // revealTooltipHandler: (data: BarData | null) => data,
    };
    const component = renderer.create(<Bar {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
