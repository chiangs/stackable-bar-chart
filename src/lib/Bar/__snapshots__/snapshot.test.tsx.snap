// Vitest Snapshot v1

exports[`Bar > Bar component renders correctly 1`] = `
<div
  className="stackable-bar stacked "
  data-testid="stackable-bar"
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  style={
    {
      "background": undefined,
      "transition": "width .2s ease-out",
      "width": "0%",
    }
  }
>
  <div
    className="bar-value"
    style={
      {
        "background": undefined,
        "transition": "width .2s ease-out",
        "width": "0%",
      }
    }
  >
    <span
      className="bar-value-text"
      style={
        {
          "color": undefined,
        }
      }
    >
      1
    </span>
  </div>
</div>
`;
